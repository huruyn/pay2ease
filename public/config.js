window.API_URL = "https://zippy-retail-api-dev.azurewebsites.net";