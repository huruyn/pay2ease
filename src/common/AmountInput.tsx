import PropTypes from 'prop-types';
import { FormControl } from 'react-bootstrap';
import {
  Dispatch,
  SetStateAction,
  useEffect,
  useState,
} from 'react';

function AmountInput({
  locale,
  currency,
  decimalPoint,
  amount,
  setAmount,
}: {
  locale: string,
  currency: string,
  decimalPoint: number
  amount: number,
  setAmount: Dispatch<SetStateAction<number>>,
}) {
  const [displayAmount, setDisplayAmount] = useState<string>('');

  const formatter = new Intl.NumberFormat(locale, {
    style: 'currency',
    currency,
    minimumFractionDigits: 0,
  });
  const symbol = formatter.format(1).replace(/\p{Number}/gu, '')[0];
  const thousandSeparator = formatter.format(11111).replace(/\p{Number}/gu, '')[1];
  const decimalSeparator = formatter.format(1.1).replace(/\p{Number}/gu, '')[1];

  const toCurrency = (number: number): string => formatter.format(number);

  const canBypass = (display: string) => {
    const lastString = display[display.length - 1];
    return display === `${symbol}0.0`
      || display === ''
      || lastString === symbol
      || lastString === symbol
      || lastString === thousandSeparator
      || lastString === decimalSeparator;
  };

  const parseLocaleNumber = (stringNumber: string) => {
    stringNumber = stringNumber === '..' ? '.' : stringNumber;
    if (canBypass(stringNumber)) {
      setDisplayAmount(stringNumber);
    } else {
      stringNumber = stringNumber.includes('.') ? stringNumber.slice(0, (stringNumber.indexOf('.')) + decimalPoint + 1) : stringNumber;
      setAmount(parseFloat(parseFloat(stringNumber
        .replace(new RegExp(`\\${symbol}`), '')
        .replace(new RegExp(`\\${thousandSeparator}`, 'g'), '')
        .replace(new RegExp(`\\${decimalSeparator}`), '.'))
        .toFixed(decimalPoint)));
      setDisplayAmount(toCurrency(parseFloat(amount.toFixed(decimalPoint))));
    }
  };

  const validate = (event: any) => {
    const charCode = (event.which) ? event.which : event.keyCode;
    if (event.target.value.includes('.') && (charCode === 110 || charCode === 190)) {
      event.preventDefault();
      return false;
    }

    if (charCode > 31
      && (charCode < 48 || charCode > 57)
      && (charCode < 96 || charCode > 105)
      && (charCode < 37 || charCode > 40)
      && charCode !== 46
      && charCode !== 110
      && charCode !== 188
      && charCode !== 190) {
      event.preventDefault();
      return false;
    }
    return true;
  };

  useEffect(() => {
    if (canBypass(displayAmount)) {
      setDisplayAmount(displayAmount);
    } else {
      setDisplayAmount(toCurrency(amount));
    }
  });

  return (
    <>
      <FormControl
        className="amount"
        placeholder="$0.00"
        value={displayAmount}
        type="string"
        onKeyDown={(evt) => validate(evt)}
        onChange={(evt) => {
          parseLocaleNumber(evt.target.value);
        }}
      />
    </>
  );
}

AmountInput.propTypes = {
  amount: PropTypes.number,
  locale: PropTypes.string,
  currency: PropTypes.string,
  decimalPoint: PropTypes.number,
};

AmountInput.defaultProps = {
  locale: 'en-CA',
  currency: 'CAD',
  decimalPoint: 2,
  amount: 0,
};

export default AmountInput;
