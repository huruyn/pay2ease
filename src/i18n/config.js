import i18n from 'i18next';
import { initReactI18next } from 'react-i18next';

i18n.use(initReactI18next).init({
  fallbackLng: 'en-US',
  lng: 'en-US',
  resources: {
    'en-CA': {
      translations: require('./locales/en-CA/translations.json')
    },
    'en-US': {
      translations: require('./locales/en-US/translations.json')
    },
    'es-US': {
      translations: require('./locales/es-US/translations.json')
    },
    'fr-CA': {
      translations: require('./locales/fr-CA/translations.json')
    },
  },
  ns: ['translations'],
  defaultNS: 'translations'
});

i18n.languages = ['en-CA', 'en-US', 'es-US', 'fr-CA'];

export default i18n;