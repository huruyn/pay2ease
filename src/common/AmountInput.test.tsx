import '../i18n/config';
import AmountInput from './AmountInput';

const ReactTestRenderer = require('react-test-renderer');

describe('AmountInput Component', () => {
  it('matches the snapshot', () => {
    const tree = ReactTestRenderer.create(
      <AmountInput
        amount={0}
        setAmount={jest.fn()}
      />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
