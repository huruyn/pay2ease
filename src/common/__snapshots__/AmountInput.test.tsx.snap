// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AmountInput Component matches the snapshot 1`] = `
<input
  className="amount form-control"
  onChange={[Function]}
  onKeyDown={[Function]}
  placeholder="$0.00"
  type="string"
  value=""
/>
`;
